### DISCUSSION TITLE
Generating test data

### DISCUSSION THREAD

**Discussion by AttilaMihaly** [finos/morphir#150] [Ideas]
# Problem Statement

During testing we often need to feed the system with some data. In some cases such as volume testing or property-based testing we don't care about the specific values only that they are valid. There are many tools out there to generate arbitrary data and they all rely on some kind of description of the shape of the data and potential constraints. Some do this with description languages, others use combinator libraries that allow you to compose generators (property-based testing frameworks tend to do this).

The Morphir IR already contains all the necessary information so we could automate this and as a Morphir user you could get this for free. 

# User Stories

- As a Morphir user, I want to generate arbitrary but valid data for specific parts of a domain model so that I can use it for volume or property-based testing.

**Comment by stephengoldbaum** [comment]
Could you elaborate on how Morphir's solution provides more than and why someone would want to choose that over the technologies you mentioned?

**Reply by AttilaMihaly** [reply]
Sure. 

1. The Morphir based solution is technology agnostic. The data that is generated here will be accessible in every technology that has integrations with Morphir already.
2. You don't have to do any extra work on top of creating your domain model. Many of the tools mentioned require you to do an additional step and compose generators that align with the types. 
3. If you already have a Morphir model you get this for free so any alternatives will require more work.

**Comment by stephengoldbaum** [comment]
So from the perspective of someone deciding whether or not to use Morphir, the advantage is that it saves you from writing custom generators?

- How does the solution know the correct results for the tests it's generated?
- If it's basing it on the code, what happens when the code changes? Does the user change the results manually or regenerate all of the tests?

**Reply by AttilaMihaly** [reply]
We are not talking about generating test cases here. It's just data for testing, mostly volume and property-based testing where you just want some valid input and don't care about the specific result.

Yes, it saves you from writing custom generators and also learning how to write them (this is not common knowledge). I think a lot of the benefit that Morphir provides is that you just need to invest once (in learning Morphir) and you get a lot of tools in many languages while normally you would have to learn each technology.

**Comment by AttilaMihaly** [comment]
Check out the corresponding proposal here: https://github.com/finos/morphir/discussions/154
